engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java

```diff
@@ -89,30 +100,101 @@ public Object convertValue(PyObject value) {
      */
     public static class CallableWrapper {
+        private static final Map<Character, Class<?>> numpyType2JavaClass = new HashMap<>();
+
+        static {
+            numpyType2JavaClass.put('i', int.class);
+            numpyType2JavaClass.put('l', long.class);
+            numpyType2JavaClass.put('h', short.class);
+            numpyType2JavaClass.put('f', float.class);
+            numpyType2JavaClass.put('d', double.class);
+            numpyType2JavaClass.put('b', byte.class);
+            numpyType2JavaClass.put('?', boolean.class);
+            numpyType2JavaClass.put('O', Object.class);
+        }
+
         private final PyObject pyObject;
 
+        private String signature = null;
+        private List<Class<?>> paramTypes;
+        private Class<?> returnType;
+        private boolean vectorizable = false;
+        private boolean vectorized = false;
+
+        // first value of the pair is null if it is a resolved constant arg, otherwise it is the used column name when
+        // uninitialized and an index value of the chunked arrays for the column sources used as args when initialized
+        private ArrayList<Pair<?, ?>> args = new ArrayList<>();
+        private Class<?>[] argTypes;
+
         public CallableWrapper(PyObject pyObject) {
             this.pyObject = pyObject;
         }
 
+        private void prepareSignature() {
+            if (pyObject.getType().equals(NUMBA_VECTORIZED_FUNC_TYPE)) {
+                List<PyObject> params = pyObject.getAttribute("types").asList();
+                if (params.isEmpty()) {
+                    throw new IllegalArgumentException(
+                            "numba vectorized function must have an explicit signature: " + pyObject);
+                }
+                // numba allows a vectorized function to have multiple signatures, only the first one
+                // will be accepted by DH
+                signature = params.get(0).getStringValue();
+                vectorized = true;
+            } else if (pyObject.hasAttribute("dh_vectorized")) {
+                signature = pyObject.getAttribute("signature").toString();
+                vectorized = true;
+            } else {
+                signature = dh_table_module.call("_encode_signature", pyObject).toString();
+                vectorized = false;
+            }
+        }
+
+        public void parseSignature() {
+            if (signature != null) {
+                return;
+            }
+
+            prepareSignature();
+
+            // the 'types' field of a vectorized function follows the pattern of '[ilhfdb?O]*->[ilhfdb?O]',
+            // eg. [ll->d] defines two int64 (long) arguments and a double return type.
+            if (signature == null || signature.length() == 0) {
+                throw new IllegalStateException("Signature should always be available.");
+            }
+
+            char numpyTypeCode = signature.charAt(signature.length() - 1);
```

Comment:
The signature is being encoded as a `String`, which does work.  Would it be cleaner or better to have a custom class that does all of the relevant manipulations?  It seems like the custom class would certainly make a tiny bit of testing easier.

Reply:
Since we immediately translate the Python signature to Java classes, the benefit of a custom class isn't very obvious. However that may change when more complex manipulation of signatures is needed in the future.